witness/src/Dscp/Snowdrop/Storage/Avlp.hs

```diff
@@ -0,0 +1,236 @@
+    <> ")"
+
+data AVLChgAccum h k v =
+  AVLChgAccum
+    { acaMap     :: AVL.Map h k v
+    -- ^ AVL map, which contains avl tree with most-recent updates
+    , acaTouched :: Set AVL.Revision
+    -- ^ Set of nodes, which were touched during all of change operations applied on tree
+    }
+
+accumFromAVLMap :: AVL.Map h k v -> AVLChgAccum h k v
+accumFromAVLMap tree = AVLChgAccum tree Set.empty
+
+initAVLStorage
+    :: forall h k v m .
+    ( AVLWrite h k v m )
+    => Map k v -> m ()
+initAVLStorage (Map.toList -> kvs) = do
+    AVL.intialiseStorageIfNotAlready @h kvs
+
+mkRocksDbAvlActions
+  :: forall h k v m n
+  .  ( MonadIO n
+     , IdSumPrefixed k
+     , AVLBase h k v m
+     )
+  => RDB.RocksDB
+  -> n (DbModifyActions (AVLChgAccum h k v) k v m (AVL.Proof h k v))
+mkRocksDbAvlActions rdb = do
+    lookups <- newTVarIO (Set.empty :: Set AVL.Revision)
+    return DbModifyActions
+        { dmaAccessActions = mkAccessActions rdb lookups
+        , dmaApply         = mkDmaApply          lookups
+        }
+  where
+    mkDmaApply lookups accum = do
+      nodesTouched <- atomically $ STM.swapTVar lookups Set.empty
+      old          <- runWrapper rdb AVL.currentRoot
+      let preproof =  nodesTouched <> acaTouched accum
+      proof        <- runWrapper rdb $ AVL.prune preproof (AVL.fullRehash old)
+
+      runWrapper rdb $ AVL.overwrite (acaMap accum)
+      return proof
+
+mkAccessActions
+  :: forall h k v m
+  .  ( IdSumPrefixed k
+     , AVLBase h k v m
+     )
+  => RDB.RocksDB
+  -> TVar (Set.Set AVL.Revision)
+  -> DbAccessActions (AVLChgAccum h k v) k v m
+mkAccessActions rdb lookups = DbAccessActions
+    { daaGetter
+    , daaModifyAccum
+    , daaIter
+    }
+  where
+    storeNodeset :: Set.Set AVL.Revision -> m ()
+    storeNodeset nodeset = do
+        atomically $ STM.modifyTVar lookups (<> nodeset)
+
+    daaGetter :: AVLChgAccum h k v -> StateR k -> m (StateP k v)
+    daaGetter accum req = do
+        let
+          getOne newAccum key = do
+            ((val, proof), _) <- runWrapper rdb $ AVL.lookup key (acaMap newAccum)
+            storeNodeset proof
+            return $ maybe Map.empty (Map.singleton key) val
+
+        ss <- mapM (getOne accum) (Set.toList req)
+        return $ Map.unions ss
+
+    daaIter
+        :: forall b
+        .  AVLChgAccum h k v
+        -> Prefix
+        -> b
+        -> ((k, v) -> b -> b)
+        -> m b
+    daaIter accum prefix start folder = do
+        (result, nodeset) <- runWrapper rdb $ AVL.foldIf (\x -> idSumPrefix x == prefix, start, folder, id) (acaMap accum)
+        storeNodeset nodeset
+        return result
+
+    daaModifyAccum
+        :: AVLChgAccum h k v
+        -> ChgAccumModifier k v
+        -> m (Either (CSMappendException k) (AVLChgAccum h k v, Undo k v))
+    daaModifyAccum oldAccum changes = do
+        case changes of
+          CAMChange cs -> do
+            applyChangeSet cs oldAccum
+
+          CAMRevert (Undo cs _wat) -> do
```

Comment:
Just accidentally saw this and I believe here is the source of a bug that undo won't be correctly applied.

Reply:
By the way what was the meaning of second `Undo` param named `_wat` here?